src/prob/pf.jl

```diff
@@ -198,3 +198,803 @@ function build_mc_pf(pm::AbstractUBFModels)
+        elseif comp_type == "transformer"
+            solution[comp_type][id] = Dict{String, Any}()
+            solution[comp_type][id]["ctrr"] = real.(c_tots)
+            solution[comp_type][id]["ctri"] = real.(c_tots)
+        end
+    end
+    solution["settings"] = deepcopy(pfd.data_math["settings"])
+    solution["per_unit"] = deepcopy(pfd.data_math["per_unit"])
+
+    return solution
+end
+
+
+# COMPONENT INTERFACES
+"""
+    _cpf_branch_interface(
+      branch::Dict,
+      v_start::Dict
+    )
+
+Branch component interface outputs branch primitive Y matrix.
+"""
+function _cpf_branch_interface(branch::Dict{String,<:Any}, v_start::Dict{<:Any,<:Any}, explicit_neutral::Bool)
+    Ys = inv(branch["br_r"]+im*branch["br_x"])
+    Yfr = branch["g_fr"]+im*branch["b_fr"]
+    Yto = branch["g_to"]+im*branch["b_to"]
+    y_prim = [(Ys.+Yfr) -Ys; -Ys (Ys.+Yto)]
+    bts = [[(branch["f_bus"], t) for t in branch["f_connections"]]..., [(branch["t_bus"], t) for t in branch["t_connections"]]...]
+    c_tots_func = function(v_bt)
+        return y_prim * v_bt
+    end
+    return  bts, 0, y_prim, nothing, c_tots_func
+end
+
+
+"""
+    _cpf_shunt_interface(
+      shunt::Dict,
+      v_start::Dict
+    )
+
+Shunt component interface outputs shunt primitive Y matrix.
+"""
+function _cpf_shunt_interface(shunt::Dict{String,<:Any}, v_start::Dict{<:Any,<:Any}, explicit_neutral::Bool)
+    Y = shunt["gs"]+im*shunt["bs"]
+    bts = [(shunt["shunt_bus"], t) for t in shunt["connections"]]
+    c_tots_func = function(v_bt)
+        return Y * v_bt
+    end
+    return  bts, 0, Y, nothing, c_tots_func
+end
+
+
+"""
+    _cpf_transformer_interface(
+      tr::Dict,
+      v_start::Dict
+    )
+
+Transformer component interface outputs transformer primitive Y matrix.
+"""
+function _cpf_transformer_interface(tr::Dict{String,<:Any}, v_start::Dict{<:Any,<:Any}, explicit_neutral::Bool)
+    f_ns = [(tr["f_bus"], t) for t in tr["f_connections"]]
+    t_ns = [(tr["t_bus"], t) for t in tr["t_connections"]]
+    ts = tr["tm_set"]*tr["tm_nom"]*tr["polarity"]
+    if tr["configuration"]==WYE
+        npairs_fr = [(f_ns[i], f_ns[end]) for i in 1:length(f_ns)-1]
+        npairs_to = [(t_ns[i], t_ns[end]) for i in 1:length(t_ns)-1]
+        bts, nr_vns, Y = _compose_yprim_banked_ideal_transformers(ts, npairs_fr, npairs_to)
+    else
+        @assert length(f_ns)==3
+        npairs_fr = [(f_ns[1], f_ns[2]), (f_ns[2], f_ns[3]), (f_ns[3], f_ns[1])]
+        npairs_to = [(t_ns[i], t_ns[end]) for i in 1:length(t_ns)-1]
+        bts, nr_vns, Y = _compose_yprim_banked_ideal_transformers(ts, npairs_fr, npairs_to)
+    end
+    c_tots_func = function(v_bt)
+        return Y * v_bt
+    end
+    return  bts, nr_vns, Y, nothing, c_tots_func
+end
+
+
+"""
+    _compose_yprim_banked_ideal_transformers(
+      ts::Vector,
+      npairs_fr::Tuple,
+      npairs_to::Tuple,
+      ppm::Float
+    )
+
+Modifies ideal transformers to avoid singularity error.
```

Comment:
"Modifies ideal transformers to avoid singularity error, through the ppm value, inspired by OpenDSS."

Reply:
updated for all three ideal transformer variants